Cubical/ZCohomology/RingStructure/GradedCommutativity.agda

```diff
@@ -0,0 +1,1082 @@
+                λ { north → cong (-ₖ-gen n m (inr p) (inr q))
+                  ; south P → cong ∣_∣ₕ (sym (merid (ptSn _))) ∙ (cong (-ₖ-gen n m (inr p) (inr q)) P)
+                  ; (merid a i) → t a i}
+    where
+    t : (a : S₊ (suc k)) → PathP (λ i → 0ₖ (2 + k) ≡ ∣ merid a i ∣ₕ → ∣ merid a i ∣ₕ ≡ 0ₖ (2 + k))
+                                  (cong (-ₖ-gen n m (inr p) (inr q)))
+                                  (λ P → cong ∣_∣ₕ (sym (merid (ptSn _))) ∙ (cong (-ₖ-gen n m (inr p) (inr q)) P))
+    t a = toPathP (funExt λ P → cong (transport (λ i → ∣ merid a i ∣ ≡ 0ₖ (suc (suc k))))
+                   (cong (cong (-ₖ-gen n m (inr p) (inr q))) (λ i → (transp (λ j → 0ₖ (suc (suc k)) ≡ ∣ merid a (~ j ∧ ~ i) ∣) i
+                         (compPath-filler P (λ j → ∣ merid a (~ j) ∣ₕ) i))))
+        ∙∙ cong (transport (λ i → ∣ merid a i ∣ ≡ 0ₖ (suc (suc k)))) (congFunct (-ₖ-gen n m (inr p) (inr q)) P (sym (cong ∣_∣ₕ (merid a))))
+        ∙∙ (λ j → transp (λ i → ∣ merid a (i ∨ j) ∣ ≡ 0ₖ (suc (suc k))) j
+                  (compPath-filler' (cong ∣_∣ₕ (sym (merid a)))
+                    (cong (-ₖ-gen n m (inr p) (inr q)) P
+                    ∙ cong (-ₖ-gen n m (inr p) (inr q)) (sym (cong ∣_∣ₕ (merid a)))) j))
+        ∙∙ (λ i → sym (cong ∣_∣ₕ (merid a))
+                 ∙ isCommΩK (2 + k) (cong (-ₖ-gen n m (inr p) (inr q)) P)
+                                    (cong (-ₖ-gen n m (inr p) (inr q)) (sym (cong ∣_∣ₕ (merid a)))) i)
+        ∙∙ (λ j → (λ i → ∣ merid a (~ i ∨ j) ∣)
+                 ∙ (cong ∣_∣ₕ (compPath-filler' (merid a) (sym (merid (ptSn _))) (~ j)) ∙ (λ i → -ₖ-gen n m (inr p) (inr q) (P i))))
+        ∙ sym (lUnit _))
+
+  code≡sym : (x : coHomK (2 + k)) → (p : 0ₖ _ ≡ x) → code x p ≡ sym p
+  code≡sym x = J (λ x p → code x p ≡ sym p) refl
+
+cong-cong-ₖ-gen-inr : {k : ℕ} (n m : ℕ) (p : _) (q : _)
+    (P : Square (refl {x = 0ₖ (suc (suc k))}) refl refl refl)
+  → cong (cong (-ₖ-gen n m (inr p) (inr q))) P ≡ sym P
+cong-cong-ₖ-gen-inr n m p q P =
+     rUnit _
+  ∙∙ (λ k → (λ i → cong-ₖ-gen-inr n m p q refl (i ∧ k))
+          ∙∙ (λ i → cong-ₖ-gen-inr n m p q (P i) k)
+          ∙∙ λ i → cong-ₖ-gen-inr n m p q refl (~ i ∧ k))
+  ∙∙ (λ k → transportRefl refl k
+          ∙∙ cong sym P
+          ∙∙ transportRefl refl k)
+  ∙∙ sym (rUnit (cong sym P))
+  ∙∙ sym (sym≡cong-sym P)
+
+Kn→ΩKn+1-ₖ' : {k : ℕ} (n m : ℕ) (p : _) (q : _) (x : coHomK k)
+  → Kn→ΩKn+1 k (-ₖ-gen n m (inr p) (inr q) x) ≡ sym (Kn→ΩKn+1 k x)
+Kn→ΩKn+1-ₖ' n m p q x = cong (Kn→ΩKn+1 _) (-ₖ-gen-inr≡-ₖ n m p q x) ∙ Kn→ΩKn+1-ₖ _ x
+
+transpΩ² : {n m : ℕ} (p q : n ≡ m) → (P : _)
+        → transport (λ i → refl {x = 0ₖ (p i)} ≡ refl {x = 0ₖ (p i)}) P
+        ≡ transport (λ i → refl {x = 0ₖ (q i)} ≡ refl {x = 0ₖ (q i)}) P
+transpΩ² p q P k = subst (λ n → refl {x = 0ₖ n} ≡ refl {x = 0ₖ n}) (isSetℕ _ _ p q k) P
+
+-- Some technical lemmas about Kn→Ω²Kn+2 and its interaction with -ₖⁿ̇*ᵐ and transports
```

Comment:
These are pretty much unreadable to me, but it looks like many cases are pretty much identical... I can't help but wonder if less pattern-matching or some lemmas could help with readability?

Reply:
Hehe, this has already been cleaned up (300 lines shorter than the original). Maybe we could just add a todo and I can go back and have a last check later. 

The problem is that you have a bunch of even and odd numbers occurring in different orders in different places. It was quite hard to capture this kind of stuff in in more general lemmas. Everything happens in Omega^2 too, so I have to rely quite a bit on reductions (hence the pattern matching).

Reply:
Ah ok! Leaving a TODO sounds good!